memory/buffer.py

```diff
@@ -0,0 +1,244 @@
+            'next_observations': self.dataset['next_observations'][indices + c - 1],  # s_c
+            'terminals': self.dataset['terminals'][indices + c - 1],  # terminal_{c-1}
+            'successes': self.dataset['successes'][indices + c - 1],  # success_{c-1}
+        }
+
+    def _preprocess(self, dataset):
+        dataset = super()._preprocess(dataset)
+        return self._split_trajectories(dataset)
+
+    def _split_trajectories(self, dataset):
+        # get subtrajectories
+        terminal_indices = np.where(dataset['terminals'])[0]
+        terminal_indices = np.insert(terminal_indices, 0, -1)  # virtual terminal before the first transition
+        if terminal_indices[-1] != len(dataset['terminals']) - 1:
+            # treat the last transition as terminal
+            terminal_indices = np.append(terminal_indices, len(dataset['terminals']) - 1)
+
+        # get starting indices of subtrajectories with length subtraj_len
+        subtraj_indices = []
+        for i in range(1, len(terminal_indices)):
+            # start_indices will be empty for too short trajectories (length < subtraj_len)
+            start_indices = range(
+                terminal_indices[i - 1] + 1,
+                terminal_indices[i] - self.subtraj_len + 2,
+                self.sliding_window_step
+            )
+            subtraj_indices += list(start_indices)
+        subtraj_indices = np.asarray(subtraj_indices)
+
+        self.terminal_indices = terminal_indices
+        self.subtraj_indices = subtraj_indices
+
+        if self.verbose:
+            traj_lens = terminal_indices[1:] - terminal_indices[:-1]
+            traj_rets = np.array(
+                [dataset['rewards'][s+1:e+1].sum() for s, e in zip(terminal_indices[:-1], terminal_indices[1:])]
+            )
+            print('[B]', '- number of trajectories: {}'.format(len(terminal_indices) - 1))
+            print('[B]', '  - average trajectory length: {:.1f} (min: {}, med: {}, max: {})'.format(
+                traj_lens.mean(), traj_lens.min(), np.median(traj_lens), traj_lens.max()))
+            print('[B]', '  - average return: {:.3f} (min: {:.2f}, med: {:.2f}, max: {:.2f})'.format(
+                traj_rets.mean(), traj_rets.min(), np.median(traj_rets), traj_rets.max()))
+            print('[B]', '- number of subtrajectories (length {}): {}'.format(
+                self.subtraj_len, len(subtraj_indices)))
+
+        return dataset
+
+
+class FixedBuffer(Dataset):
+    def __init__(self, data_dict, preproc={}):
+        self.data_dict = data_dict
+        self.preproc = {k: (lambda x: x) for k in data_dict}
+        self.preproc.update(preproc)
+        self.length = len(next(iter(self.data_dict.values())))
+
+    def __getitem__(self, idx):
+        return {k: self.preproc[k](v[idx]) for k, v in self.data_dict.items()}
+
+    def __len__(self):
+        return self.length
+
+
+class RandomSamplingWrapper:
```

Comment:
Why is this wrapper necessary? Couldn't it be done by just adding a `random_sample` method in the Buffer class?

Reply:
It was mainly due to the different `__len__`. I could add random-related options to the buffer, but the wrapper seems to be the clearer way.